types/src/v0/impls/block/full_payload/ns_proof/avidm.rs

```diff
@@ -0,0 +1,192 @@
+            return None; // error: index out of bounds
+        }
+
+        if ns_table[index].is_empty() {
+            None
+        } else {
+            match NsAvidMScheme::namespace_proof(common, &payload.raw_payload, index, ns_table) {
+                Ok(proof) => Some(AvidMNsProof(proof)),
+                Err(e) => {
+                    tracing::error!("error generating namespace proof: {:?}", e);
+                    None
+                },
+            }
+        }
+    }
+
+    /// Unlike the ADVZ scheme, this function won't fail with a wrong `ns_table`.
+    /// It only uses `ns_table` to get the namespace id.
+    pub fn verify(
+        &self,
+        ns_table: &NsTable,
+        commit: &VidCommitment,
+        common: &AvidMCommon,
+    ) -> Option<(Vec<Transaction>, NamespaceId)> {
+        match commit {
+            VidCommitment::V1(commit) => {
+                match NsAvidMScheme::verify_namespace_proof(common, commit, &self.0) {
+                    Ok(Ok(_)) => {
+                        let ns_id = ns_table.read_ns_id(&NsIndex(self.0.ns_index))?;
+                        let ns_payload = NsPayload::from_bytes_slice(&self.0.ns_payload);
+                        Some((ns_payload.export_all_txs(&ns_id), ns_id))
+                    },
+                    Ok(Err(_)) => None,
+                    Err(e) => {
+                        tracing::warn!("error verifying namespace proof: {:?}", e);
+                        None
+                    },
+                }
+            },
+            _ => None,
+        }
+    }
+}
+
+/// Copied from ADVZNsProof tests.
+#[cfg(test)]
+mod tests {
+    use futures::future;
+    use hotshot::{helpers::initialize_logging, traits::BlockPayload};
+    use hotshot_types::{
+        data::VidCommitment,
+        traits::EncodeBytes,
+        vid::avidm::{AvidMParam, AvidMScheme},
+    };
+
+    use crate::{v0::impls::block::test::ValidTest, v0_3::AvidMNsProof, NsIndex, Payload};
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn ns_proof() {
+        let test_cases = vec![
+            vec![
+                vec![5, 8, 8],
+                vec![7, 9, 11],
+                vec![10, 5, 8],
+                vec![7, 8, 9],
+                vec![],
+            ],
+            vec![vec![1, 2, 3], vec![4, 5, 6]],
+            vec![],
+        ];
+
+        initialize_logging();
+
+        let mut rng = jf_utils::test_rng();
+        let mut tests = ValidTest::many_from_tx_lengths(test_cases, &mut rng);
+
+        let param = AvidMParam::new(5usize, 10usize).unwrap();
+
+        struct BlockInfo {
+            block: Payload,
+            vid_commit: VidCommitment,
+            ns_proofs: Vec<AvidMNsProof>,
+        }
+
+        let blocks: Vec<BlockInfo> = {
+            // compute blocks separately to avoid async error `captured variable
+            // cannot escape `FnMut` closure body` caused by mutable variable `vid`
```

Comment:
the mutable variable is commented out now though

Reply:
restructured in [97a33eb](https://github.com/EspressoSystems/espresso-network/pull/2866/commits/97a33eb2939a5d86c78b7543f1c9145c2a1897f5)